[4.0] Replace chosen.js to choice.js

Fedik opened the pull request:
### Summary of Changes
This a try to to replace chosen.js to choice.js.

- [x] Module position field 
- [x] Tags field
- [x] Category field
- [x] Filter fields

@wilsonge is any else field is missed in the list?

#### Testing:

Module editing: create custom position, search existing positions
Article editing: test Tag field, test category field
Articles filter: test the filter by multiple category, Tag, etc, 

Everything should work as before the patch.

### Documentation Changes Required
maybe

Fedik committed:
add choice-js

Fedik committed:
FieldModuelEditposition custom element

Fedik committed:
FieldModuelEditposition translation

Fedik committed:
FieldModuelEditposition extend FormField is enough

Fedik committed:
Sync. Merge branch '4.0-dev' into choice-js

 Conflicts:
	administrator/components/com_modules/tmpl/module/edit_positions.php

Fedik committed:
field-tags

Fedik committed:
field-tags make custom term to work

brianteeman commented on administrator/components/com_modules/forms/module.xml:
```diff
@@ -78,71 +78,71 @@
-			type="moduleposition"
+		<field
+			name="position"
+			type="ModulesPosition"
```
didnt we have an issue with case in the xml files before?

Fedik replied:
it because class autoloader, 
if I keep `modulesposition` it will try to search for `ModulespositionField.php` but actual file name is `ModulesPositionField.php`

Fedik replied:
yes, there was an issue about it, but cannot find

Fedik committed:
FieldModuleEditposition: fix typoe and cs

Fedik committed:
FieldModuleEditposition: remove unused stuff

Fedik committed:
field-tags request for remote tags

Fedik committed:
field-tags deprecate some stuff

Fedik committed:
Merge branch 'choice-js-tags' into choice-js

brianteeman commented on build/build-modules-js/settings.json:
```diff
@@ -116,6 +121,19 @@
         "licenseFilename": "LICENSE"
       },
+      "choices.js": {
+        "name": "choices.js",
```
This need to be changed to just choices otherwise you get this error in the build

`Error: ENOENT: no such file or directory, open 'C:\htdocs\40\media\vendor\choices.min.js\css\choices.css'
`

Fedik replied:
That what I am had afraid of. Currently I am working with "dev" mode and on on Linux, so all works fine.
The problem that the package name is `choices.js` https://github.com/jshjohnson/Choices/blob/master/package.json

This will be need to find a solution.

brianteeman replied:
So the problem is with our build scripts and need to be solved there or it will happen again with another package (I'm still new to npm)

wilsonge replied:
You should be ok changing the name on this line to choices. The key on line 124 is the path to the node_modules dir

Fedik replied:
> The key on line 124 is the path to the node_modules dir

@wilsonge well, no, I thought the same, I tried, and that not work.
but I think it is fine not keep an original name, 
just have to find out why an Install task fail on Windows

I think I have fixed `Error: ENOENT: no such file or directory,`, there was incorrect replacing of the file extension.
But @brianteeman said there now some empty CSS file, after build, and I do not have a windows to check in details.

I have looked in my windows but there only a trees :smile:

wilsonge replied:
![image](https://user-images.githubusercontent.com/1986000/45931225-e1c0a880-bf62-11e8-9e94-e29f8dd6fd18.png)

Well that definitely still builds ok and drops the `.js` in the folder path. I don't have windows either and Rowan who's my go to is on holiday at the moment. @brianteeman can I try sorting out a screenshare with you at some point this week after work?

Fedik replied:
hm, okay, then I missed something :smiley:

brianteeman replied:
@wilsonge of course just name a time/date

dgrammatiko replied:
@wilsonge @Fedik the name part in the json is supposed to set a custom name for the vendor. IIRC there are couple of crappy names in the dependencies with @ or other strange .something parts in the names. If the code is not doing what is supposed to do just blame it on me but fwiw it was implemented based on some specific cases...

Fedik replied:
I have renamed to `choicesjs`

Fedik committed:
choice-js-cat categoriedit field

Fedik committed:
choice-js-cat categoriedit make it work

Quy commented on administrator/components/com_categories/layouts/joomla/form/field/categoryedit.php:
```diff
@@ -0,0 +1,139 @@
+defined('JPATH_BASE') or die;
+
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\HTML\HTMLHelper;
```
Sort in alpha order

Quy commented on administrator/components/com_categories/layouts/joomla/form/field/categoryedit.php:
```diff
@@ -0,0 +1,139 @@
+$attr .= $required ? ' required' : '';
+$attr .= $autofocus ? ' autofocus' : '';
+
+// To avoid user's confusion, readonly="true" should imply disabled="true".
```
Change to `disabled="disabled"`?

Quy commented on administrator/components/com_categories/layouts/joomla/form/field/categoryedit.php:
```diff
@@ -0,0 +1,139 @@
+?>
+
+<joomla-field-categoryedit><?php
+	echo implode($html);
```
Make it one line `<?php echo implode($html); ?>`

Fedik replied:
you are looking in wrong commit :smiley:

Quy commented on administrator/components/com_categories/layouts/joomla/form/field/categoryedit.php:
```diff
@@ -0,0 +1,139 @@
+
+Text::script('JGLOBAL_SELECT_NO_RESULTS_MATCH');
+Text::script('JGLOBAL_SELECT_PRESS_TO_SELECT');
+//Text::script('JGLOBAL_TYPE_OR_SELECT_SOME_OPTIONS');
```
Delete comment?

Fedik replied:
it not finished, anything can happen :neckbeard:

Fedik committed:
merge JS to single field

Fedik committed:
prevent from submit on "enter"

Fedik committed:
add some extar info

Fedik committed:
add layout to ListField

Fedik committed:
joomla.form.field.list-fancy-select layout for list field

Fedik committed:
joomla.form.field.accesslevel for accesslevel field

Fedik committed:
joomla.form.field.accesslevel-fancy-select for accesslevel field

Fedik committed:
update multiselect fields in Articles filter

Fedik committed:
precache existing options

Fedik committed:
update multiselect fields in Featured filter

Fedik committed:
fix filters, ignore click inside the filter container

Fedik commented on build/media_src/system/js/searchtools.es6.js:
```diff
@@ -179,13 +179,28 @@
         if (document.body.classList.contains('filters-shown')) {
+
+          // Ignore click inside the filter container
+          if (event.composedPath && typeof(event.composedPath) === 'function') {
```
whot?

Fedik commented:
@brianteeman can you please tell when exactly you got that error?  `no such file or directory, open 'C:\htdocs\40\media\vendor\choices.min.js\css\choices.css`
Maybe some screenshot.

I tried to build a couple times and cannot reproduce it currently.

Fedik committed:
phpcs and hound

brianteeman commented:
It was part of the output of the npm i
I am on a train on a way to a sprint but as soon as I get there I will check this and provide more info

Fedik committed:
revert ModulesPosition and create ModulesPositionedit for edit action

brianteeman commented:
```processing: C:\htdocs\40\media\vendor\choices.js\css\choices.css
fs.js:145
    throw err;
    ^

Error: ENOENT: no such file or directory, open 'C:\htdocs\40\media\vendor\choices.min.js\css\choices.css'
    at Object.fs.openSync (fs.js:579:3)
    at Object.fs.writeFileSync (fs.js:1309:33)
    at files.forEach (C:\htdocs\40\build\build-modules-js\update.js:302:9)
    at Array.forEach (<anonymous>)
    at uglifyLegacyFiles (C:\htdocs\40\build\build-modules-js\update.js:294:9)
    at Object.copyAssets (C:\htdocs\40\build\build-modules-js\update.js:329:11)
    at Object.<anonymous> (C:\htdocs\40\build.js:65:22)
    at Module._compile (internal/modules/cjs/loader.js:702:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:713:10)
    at Module.load (internal/modules/cjs/loader.js:612:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:551:12)
    at Function.Module._load (internal/modules/cjs/loader.js:543:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:744:10)
    at startup (internal/bootstrap/node.js:238:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:572:3)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! joomla@4.0.0 install: `node build.js --copy-assets && node build.js --build-check`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the joomla@4.0.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:```

![image](https://user-images.githubusercontent.com/1296369/45916076-da60a880-be57-11e8-805f-56e182fe03a1.png)

Fedik commented:
@wilsonge I need to load the field from com_modules, in Module frontend editing form `/components/com_config/forms/modules.xml` 
I am doing with `addfieldprefix`:
```
<field
	name="position"
	type="ModulesPositionedit"
	label="COM_MODULES_FIELD_POSITION_LABEL"
	addfieldprefix="Joomla\Component\Modules\Administrator\Field"
/>
```
but nothing works, what is wrong?

Fedik committed:
Frontend module editing

Fedik committed:
Fix styling

Fedik committed:
And forntend, fix styling

Fedik committed:
hound

Fedik committed:
try to fix build/update.js for Windows

Fedik commented:
@brianteeman please try one more run, 
I think I found a reason, and made a fix, but cannot test my own

brianteeman commented:
Sorry that change didnt work

Buiild with the latest change produced a css file with the content just the word "undefined" and a blank minified css file

Fedik commented:
okay, then that error fixed :wink:  

now, what is file path contain "undefined"?

Fedik committed:
remove lasts uses of 'formbehavior.chosen'

brianteeman commented:
C:\htdocs\40\media\vendor\choices.js\css\choices.css

Fedik commented:
I think it ready to test, if there someone brave enough :neckbeard: 

Would be good if someone who have a Windows, to look at build failure, under Windows.
I cannot find why `\media\vendor\choices.js\css\choices.css` become empty after build.

dgrammatiko commented on build/media/webcomponents/js/field-fancy-select/field-fancy-select.js:
```diff
@@ -0,0 +1,178 @@
+    get searchPlaceholder() { return this.getAttribute('search-placeholder'); }
+
+    connectedCallback() {
+      if (!window.Choices) {
```
@Fedik it would be way better if you inject the choices.js in this file rather than expecting the script to live in the window (but that's maybe my personal preference here so you can ignore it)

Fedik replied:
I not very like to inject external library :wink:

dgrammatiko replied:
That’s fine as joomla is loading all the custom elements on DOM loaded event. From a standards point of view custom elements should be able to load async which obviously dependencies won’t be allowed as you cannot ensure the order of the scripts.FWIW check the way I did the loading of code mirror in the relative custom element, I guess that is closer to the specs. But let me write this again: joomla is loading custom elements on dom complete so this is more of a future aware thing, not something that will affect the current stack

Fedik replied:
better I would not looked there :smile: 
no, that no, if the script will never be loaded it is equal `while(true){}`, but with pauses per frame   :wink: 
there is `window.addEventListener('load' ...)` as last rescue

dgrammatiko replied:
Well, I never said that it was a solid implementation 😂 but you get the idea 💡

wilsonge commented:
OK So I just did a quick test here and things seems to work for module positions, tags and categories. There is a change here in that before we used to do ajax searching on tags after 3 characters were entered - now we are just loading all the tags in the system. But this is looking good. We obviously need to get windows sorted

Fedik commented on libraries/cms/html/formbehavior.php:
```diff
@@ -17,7 +17,9 @@
- * @since  3.0
+ * @since       3.0
+ *
+ * @deprecated  Without replacement
```
@wilsonge please check, how to deprecate it correctly

Fedik commented on libraries/cms/html/tag.php:
```diff
@@ -165,6 +165,8 @@ public static function tags($config = array('filter.published' => array(0, 1)))
 	 *
 	 * @since   3.1
+	 *
+	 * @deprecated  Without replacement
```
@wilsonge and here also, how to deprecate it correctly

wilsonge replied:
It just needs a version after the `@deprecated` tag (5.0) before the text

Fedik replied:
and can we deprecate it for J3 and remove it from J4 or better keep? :smiley:

Fedik committed:
rename choices.js to choicesjs

Fedik committed:
Improve Tags ajax search

Quy commented on libraries/src/Form/Field/TagField.php:
```diff
@@ -263,11 +276,28 @@ public function isNested()
+	/**
+	 * Check whether need to enable AJAX search
+	 *
+	 * @return  bool
```
spell out `boolean`

Fedik committed:
boolean

laoneo commented:
Can we get here some tests? Thanks

brianteeman commented:
@laoneo it doesnt work in windows - waiting for @wilsonge to use my windows computer

wilsonge committed:
Merge branch '4.0-dev' into choice-js

Fedik committed:
Update deprecated tag

laoneo commented:
Could you guys sort out the windows issue?

brianteeman commented:
I am still here waiting for @wilsonge to use my computer

wilsonge committed:
Merge branch '4.0-dev' into choice-js

wilsonge commented:
I'm off work today. Around till 5pm. Pinged you on twitter

wilsonge commented:
Spent some time today with Brian. He should be submitting a PR later

brianteeman commented:
For reference the problem is **not** with this PR but the build scripts on  windows. 

That will be a seperate PR I will submit tonight and doesnt really prevent this one fromm being merged

brianteeman commented:
I am getting this error when opening an article with some fields

`Warning: DOMDocument::loadHTML(): Tag joomla-field-fancy-select invalid in Entity, line: 2 in C:\htdocs\40\administrator\components\com_fields\Helper\FieldsHelper.php on line 315
`

brianteeman commented:
see #22706 for the corrected(?) build script

Fedik commented:
I will check this weekend

brianteeman commented:
I "think" the error is related to the calendar field

Fedik committed:
Sync. Merge branch '4.0-dev' into choice-js

 Conflicts:
	administrator/components/com_categories/Field/CategoryeditField.php
	administrator/components/com_modules/tmpl/module/edit_positions.php
	build/build-modules-js/update.js

Fedik commented:
> I am getting this error when opening an article with some fields

@brianteeman please tell me which field types do you have in that article?

Fedik commented:
@laoneo do you remember what this thing is doing? 
https://github.com/joomla/joomla-cms/blob/0e641a302f8305191011134396289e3c40d6187e/administrator/components/com_fields/Helper/FieldsHelper.php#L314-L316

the warning because DOMDocument not know about a custom tags, hm

brianteeman commented:
There were two fields. Calendar and text

Fedik committed:
DOMDocument::loadHTML and CE

Fedik committed:
choices.js and WebAsset

Fedik committed:
get/set value, and required validation

Fedik commented:
@brianteeman should work now

HLeithner commented on administrator/components/com_fields/Helper/FieldsHelper.php:
```diff
@@ -312,7 +312,10 @@ public static function prepareForm($context, Form $form, $data)
 			// Choose the first category available
 			$xml = new \DOMDocument;
+			libxml_use_internal_errors(true);
 			$xml->loadHTML($formField->__get('input'));
```
Sry if its already answered but what happens here that this hack is needed?

Fedik replied:
see https://github.com/joomla/joomla-cms/pull/22263#issuecomment-431077192

DOMDocument shows a warning on unknown elements such as html5 and custom elements

HLeithner replied:
Ok thx

brianteeman commented:
@Fedik yes that problem has been resolved - thanks

wilsonge commented:
So just to be clear is this PR now working (with a bit of a grim hack to the DomDocument stuff?)

Fedik commented:
for me, working :)

wilsonge commented:
Thankyou for this!
